src/waveform/renderers/allshader/waveformrendererpreroll.cpp

```diff
@@ -125,36 +151,20 @@ void WaveformRendererPreroll::paintGL() {
         // has changed size last time.
         m_markerLength = markerLength;
         m_markerBreadth = markerBreadth;
-        m_texture.setData(drawPrerollImage(m_markerLength,
-                m_markerBreadth,
-                m_waveformRenderer->getDevicePixelRatio(),
-                m_color));
+        dynamic_cast<PatternMaterial&>(material())
+                .setTexture(std::make_unique<Texture>(m_waveformRenderer->getContext(),
+                        drawPrerollImage(m_markerLength,
```

Comment:
does this allocate continuously in the hot `paintGL` path?

Reply:
No, only if you change the zoom during preroll.

Reply:
Because it needs to recompute the texture due to the changed zoom level?

Reply:
Correct.